design/kata-2-0-metrics.md

```diff
@@ -0,0 +1,1343 @@
+# Kata 2.0 Metrics Design
+
+* [Limitations of Kata 1.x and the target of Kata 2.0](#limitations-of-kata-1x-and-the-target-of-kata-20)
+* [Metrics architecture](#metrics-architecture)
+  * [Kata magent](#kata-magent)
+  * [Kata runtime](#kata-runtime)
+  * [Kata agent](#kata-agent)
+  * [Performance and overhead](#performance-and-overhead)
+* [Metrics list](#metrics-list)
+  * [Kata agent metrics](#kata-agent-metrics)
+  * [Kata guest OS](#kata-guest-os)
+  * [Hypervisor metrics](#hypervisor-metrics)
+  * [Kata magent metrics](#kata-magent-metrics)
+  * [Kata shim metrics](#kata-shim-metrics)
+
+
+Kata implement CRI's API and support [`ContainerStats`](https://github.com/kubernetes/kubernetes/blob/release-1.18/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto#L101) and [`ListContainerStats`](https://github.com/kubernetes/kubernetes/blob/release-1.18/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto#L103) interfaces to expose containers metrics. User can use these interface to get basic metrics about container.
+
+But unlike `runc`, Kata is a VM-based runtime and has a distinguish architecture.
+
+## Limitations of Kata 1.x and the target of Kata 2.0
+
+There are some limitations, or called missing parts, the lack of observability may become an obstacles to run Kata containers in large scale.
+
+In Kata 2.0, these parts will be took to provide more details about the system.
+
+- Shim process(Kata runtime)
+- Hypervisor stats
+- Agent process
+- Guest OS stats
+
+## Metrics architecture
+
+Kata 2.0 metrics strongly depend on [Prometheus](https://prometheus.io/), a graduated project from CNCF.
+
+The new design will introduce a new command(component) named `kata-magent`, the `m` of `magent` means metrics or management, this command is used to manage/monitor Kata runtime it self other than containers. It's the management interface with Kata runtime, and we can do something like these:
+
+- Get metrics
+- Get events
+- Extend hypervisor commands
+
+In this document we will cover metrics only. And until now it only supports metrics function.
+
+This is the architecture overview metrics in Kata Containers 2.0.
+
+![Kata Containers 2.0 metrics](arch-images/kata-2-metrics.png)
+
+
+And the sequence diagram is shown below:
+
+![Kata Containers 2.0 metrics ](arch-images/kata-metrics-sequence-diagram.png)
+
+
+### Kata magent
+
+`magent` is a management agent on one nodes, where many Kata containers are running. `magent`'s work include:
+
+- Aggregate sandbox metrics running on this node, and add `sandbox_id` label
+- As a prometheus target, all metrics from Kata shim on this node will be collected by prometheus indirectly. This can easy the targets count in Prometheus, and also need not to expose shim's metrics by `ip:port`
+
+Only one `magent` process are running on one node.
```

Comment:
Could you add some details about security here:

- Who can connect to this service? `root` only?
- Is there any further access control aside from which user is allowed to collect? Can a non-priv user for example gather *all* stats for all containers? I'm assuming not as that would potentially expose sensitive system / workload information to the user.

Random other thoughts:

- Is this service started automatically? By systemd?
- Are there any issues running this service with https://github.com/kata-containers/ksm-throttler?

Reply:
- Metrics service is a network service, normally anyone can access it with if network is accessible.
- Regarding that all hosts running Kata Containers are not exposed to internet directly,  and access control can done by network policies. 
- If a user can access metrics service, he/she can get all metrics, there are no fine grained access control now. 
- But in the further we cant integrate  authentication/authorization if there are requirements.
- Metrics service is focus on system view point, and only for OPS, there are nothing about business data or sensitive information.

For two other comments:

- I will create a `systemd` unit file to manage metrics service.
- There will no issues with `ksm-throttler`.

Reply:
This is great, thanks @liubin.  Thinking of K8S -- do you think this should be available as a deployment via daemonset?

Reply:
Yes, that will make running metrics service more easier. I will try to add it.